Initial C.I. based on Github Actions

dmacvicar opened the pull request:
Inspired by [this comment](https://github.com/cycfi/elements/issues/70#issuecomment-597356163).

Some design notes:

* Builds on Ubuntu, Windows with Visual Studio, and MacOS.
* Tries to use platform packaged software where possible (apt-get, choco, brew)
* When not sufficient, uses the Github runner installed software
* For now, only builds and shows a badge on the README

I am sure lot of things can be improved or added, but it is a first step. Feedback appreciated!.

- 673c973 fixes a build error I got while getting all platforms to build

djowel commented on examples/typography/CMakeLists.txt:
```diff
@@ -10,8 +10,7 @@ set(ELEMENTS_APP_ID "com.cycfi.typography")
 set(ELEMENTS_APP_VERSION "1.0")
 
 set(ELEMENTS_APP_RESOURCES
-   ${CMAKE_CURRENT_SOURCE_DIR}/resources/OpenSansCondensed-Light.ttf
```
I suppose this is removed because it's not used in the code?

djowel commented:
> * [673c973](https://github.com/cycfi/elements/commit/673c973fadeaf02715ba317b87111e6c3cfe1692) fixes a build error I got while getting all platforms to build

Ahh! This. Hmmm... What build error? Oh, is it not in the font resources?

djowel commented:
Wow, I know nothing about CI. I was planning to invest some time learning about at, but looking at the code, you saved a lot of tome for me. Much appreciated!

Now I can add some real tests! This is the only library that I authored without tests from the start. GUI libs are tricky to test. 

BTW, I am currently decoupling the 2d-graphics back-end, and targetting native host-native facilities whenever possible, and also skia. The WIP repository is here:

https://github.com/cycfi/artist. 

It is named "artist" and intends to be a "High performance, cross Platform C++ Vector Graphics Library, inspired by the HTML5 canvas API." I am indeed getting good results with skia even at high-DPI, but native quartz-2d is also impressive. I started this test-driven. Would you be interested in getting that CI'd too? Right now, it's MacOS only, but it's intended to be cross-platform. The intent is that "artist:" will replace the cairo based 2d "canvas" support in elements.

dmacvicar commented:
> > * [673c973](https://github.com/cycfi/elements/commit/673c973fadeaf02715ba317b87111e6c3cfe1692) fixes a build error I got while getting all platforms to build
> 
> Ahh! This. Hmmm... What build error? Oh, is it not in the font resources?

One font was added twice. One in the top build file, one time in the specific example. This made ninja detect two overlapping targets.

dmacvicar commented:
> Wow, I know nothing about CI. I was planning to invest some time learning about at, but looking at the code, you saved me a lot of time. Much appreciated!

Thanks. While I have experience with CI and very complex test scenarios, I had only played with Github actions ince when creating a pipeline that uses emacs to generate my website (https://duncan.codes).

I was shopping for a GUI library for a personal project and I started using [nanogui](https://github.com/mitsuba-renderer/nanogui), which was pretty good, but the fact that _elements_ is also used for audio tools, it makes it a much better JUCE alternative than nanogui, and already has things like knobs. It looks much better in general.

I still have some philosophical differences with the world of boost c++. I used it extensively while working in [ZYpp](https://github.com/openSUSE/libzypp) years ago, I never got into the template meta-programming mindset, and I tend to click easier with the C++ style featured in project like like [Godot](https://github.com/godotengine/godot) or [SerenityOS](https://github.com/SerenityOS/serenity) vs boost.

One of the first things that almost drove me away from using _elements_ was the use of dependencies (json, infra, boost) for very little things, but I think I can live with that for how it looks and feels and some other things can be improved (I mentioned cmake _FetchContents_, which I already implemented, but preferred to contribute things one by one).

> Now I can add some real tests! This is the only library that I authored without tests from the start. GUI libs are tricky to test.

If I am lucky integrating _elements_ into my toy, I may also feel in the need to help here.

> BTW, I am currently decoupling the 2d-graphics back-end, and targeting host-native facilities whenever possible, and also skia. The WIP repository is here:
> 
> https://github.com/cycfi/artist.
> 
> It is named "artist" and intends to be a "High performance, cross Platform C++ Vector Graphics Library, inspired by the HTML5 canvas API." I am indeed getting good results with skia even at high-DPI, but native quartz-2d is also impressive. I started this test-driven. Would you be interested in getting that CI'd too? Right now, it's MacOS only, but it's intended to be cross-platform. The intent is that "artist:" will replace the cairo based 2d "canvas" support in elements.

I will have a look. I don't have access to Mac hardware, and Github actions require to push the CI code in order to see if it works, so the more multi-platform, the easier it is for me to test something locally first. The workflow I contributed is the squashed result of +70 commits I pushed and fixed after seeing it fails because of wrong boost versions, not having experience with Visual Studio building process and tens of other things (I learned quite a bit too).

Just curious, what was the reason you went with skia vs eg. wrapping nanovg?

djowel commented:
> Just curious, what was the reason you went with skia vs eg. wrapping nanovg?

You know that nanovg is the first vector-2d toolkit I used. I switched it out early because of poor text rendering quality and inferior rendering quality (in general) compared to my gold standard: quartz-2d (For example, nanovg does not have true shadows, cairo does not have shadows as well, but I was led to believe it was in the pipeline since it was in their roadmap, but there's no real action in the cairo stage now). This article shared my issues with text rendering quality: https://blog.johnnovak.net/2016/05/29/cross-platform-gui-trainwreck-2016-edition/#enter-nanovg . NanoVG text handling is a hack. Text nandling is VERY difficult. Even Skia delegates "text shaping" to other libraries such as Pango and HarfBuzz.

Another is lackluster support. I posted an issue in 2016 which was never acted on. Every now and then, people bring back the same issue, again text handling hacks: https://github.com/memononen/nanovg/issues/281

djowel commented:
> One of the first things that almost drove me away from using _elements_ was the use of dependencies (json, infra, boost) for very little things, but I think I can live with that for how it looks and feels and some other things can be improved (I mentioned cmake _FetchContents_, which I already implemented, but preferred to contribute things one by one).

Infra and Json are ultra-light, header only libraries. Elements use Boost asio though for handling asynchronous tasks, and asio has splendid multithreading support. Surely there are other inevitable big dependencies too, such as cairo, which also depends on other libraries down the line. 

While libraries such as nonvg look sexy/slim without such dependencies, once you get past a certain level of code maturity, you will realize that what's there is not sufficient (e.g. the hacky text shaping and rendering of nanovg). Skia has lots of other dependencies too, unfortunately more than cairo :(... 

But perhaps the issue here is how to deal with these dependencies in a way that still makes it easy to use (building skia is a nightmare that I wouldn't want the user to endure, so instead I am taking the prebuilt-binaries route).

Tell me about your gripe against dependencies and let's talk about it some more.

djowel commented:
> I still have some philosophical differences with the world of boost c++. I used it extensively while working in [ZYpp](https://github.com/openSUSE/libzypp) years ago, I never got into the template meta-programming mindset, and I tend to click easier with the C++ style featured in project like like [Godot](https://github.com/godotengine/godot) or [SerenityOS](https://github.com/SerenityOS/serenity) vs boost.

I authored a few Boost libraries and I must admit I am a template nerd too.  I tend to lean towards easily maintainable and easier to use interfaces now, but I still use template tricks in implementation 1) If it helps give a better user experience (e.g. elements composition) and 2) performance.

Tell me about the C++ style featured in project like Godot, perhaps we can meet half-way. Tell me about your philosophical differences with the world of boost c++. I am sure you know the c++ standard and even STL is based on things like functional and generic programming. Are you disliking this 'multi paradigm' approach? I sincerely want to know (and learn).

dmacvicar commented:
> Tell me about your gripe against dependencies and let's talk about it some more.

Coming from a Linux world where the whole software is packaged from source, built without internet access, and at the same time, as a developer I want to pass the Joel (Spolsky) test "can I make a build with a simple command" (which in turns simplifies C.I. a lot).

This can be achieved being careful. Looking for dependencies with some defined order and the right fallbacks (eg. system pkg-config, specified location, go to the internet) and override possibilities.

Skia having dependencies is ok, as long as it is possible to satisfy a one click build that can be customized. Lets see how hard it is. _servo_ and _chromium_ vendoring skia makes me think it is not.

djowel commented:
> Coming from a Linux world where the whole software is packaged from source, built without internet access, and at the same time, as a developer I want to pass the Joel (Spolsky) test "can I make a build with a simple command" (which in turns simplifies C.I. a lot).

That is indeed a good goal! I am all for it. I think it can be achieved. I'll think about it some more. For example, I've been mulling around the idea of a super-project that is basically an umbrella that includes the dependencies as sub modules. Right now, I am using cmake's ExternalProject_Add for skia and a prebuilt binary that I packaged and ExternalProject_Add does all the downloading, unpacking and installing in the build directory. It works very smoothly. I can do the same for Boost by decoupling the required libraries, and packaging them as modular things (I've done this before and it works fine).

> Skia having dependencies is ok, as long as it is possible to satisfy a one click build that can be customized. Lets see how hard it is. servo and chromium vendoring skia makes me think it is not.

OK, agreed! A one click build is a worthy goal.

dmacvicar commented:
> I authored a few Boost libraries and I must admit I am a template nerd too. I tend to lean towards easily maintainable and easier to use interfaces now, but I still use template tricks in implementation 1) If it helps give a better user experience (e.g. elements composition) and 2) performance.
> 
> Tell me about the C++ style featured in project like Godot, perhaps we can meet half-way. Tell me about your philosophical differences with the world of boost c++. I am sure you know the c++ standard and even STL is based on things like functional and generic programming. Are you disliking this 'multi paradigm' approach? I sincerely want to know (and learn).

I think the API you are aiming with _elements_ looks very nice and clean to my eyes.
I still need to test myself how it looks on the screen once the compiler gives you an error :-)

Regarding those projects, it is hard to describe it exactly. Like everything, the point is in the balance between clever and simple. C++ is a language packed with features and tricks, which means there is more room too much in both directions.
From the contributors perspective, not every C++ developer masters all those tricks. Being able to contribute to a code-base overloaded with those becomes just out of reach.
Yes, I am aware that std uses lot of this. I haven't used boost for a while. Perhaps my memory is just exaggerating things.

dmacvicar commented:
> That is indeed a good goal! I am all for it. I think it can be achieved. I'll think about it some more. For example, I've been mulling around the idea of a super-project that is basically an umbrella that includes the dependencies as sub modules. Right now, I am using cmake's ExternalProject_Add for skia and a prebuilt binary that I packaged and ExternalProject_Add does all the downloading, unpacking and installing in the build directory. It works very smoothly. I can do the same for Boost by decoupling the required libraries, and packaging them as modular things (I've done this before and it works fine).

I think something like that can be enough. I think I tried this in my fork with infra/json and then reverted it. Can't remember why.
With sumodules, LInux packagers have the option of making a tarball including all the recursive submodules, and then deleting those provided by the base system just before buildng, by using overrides provided by the build system (eg. the build system doing pkg_config find if the library is not in external/ submodule.

I think build system improvements can be done as CI is put in place.

djowel commented:
> I think the API you are aiming with _elements_ looks very nice and clean to my eyes.
> I still need to test myself how it looks on the screen once the compiler gives you an error :-)

Ah the dreaded c++ error message. I still struggle with it even after 20 years of template coding. Hopefully, c++20 concepts will straighten things up. The main issue is that templates are duck-typed beasts. There are a lot more C++ facilities now that make it easier and give better user experience. I am very eager to give that balanced user experience with your (as well as other) feedback.

> From the contributors perspective, not every C++ developer masters all those tricks. Being able to contribute to a code-base overloaded with those becomes just out of reach.

Nods... That tends to be forgotten.

Xeverous commented:
> The main issue is that templates are duck-typed beasts. There are a lot more C++ facilities now that make it easier and give better user experience.

We have already added some static assertions. Some of them turned out to be too restrictive and had to be removed. But I still think there is a lot of assumptions about many `T`s in elements that could be handled well without concepts.

djowel commented:
> > The main issue is that templates are duck-typed beasts. There are a lot more C++ facilities now that make it easier and give better user experience.
> 
> We have already added some static assertions. Some of them turned out to be too restrictive and had to be removed. But I still think there is a lot of assumptions about many `T`s in elements that could be handled well without concepts.

There is no substitute for real concepts. It's not really just about asserting preconditions, although that can indeed be a good start. What's more important IMO are the allowed expressions. It just so happens that elements make use of classes that already constrain the allowed expressions via inheritance, but there are alternative design and implementation strategies that could benefit from true concepts.
